Testng-eclipse can not find classes in non exported packages in a java modular maven project

### Problem Statement
Using a minimal standard java modular maven project with separate src and test folders runs as expected as long as the class beeing tested is in a package that is exported in modular-info.java. If it is not, testng-eclipse can not find the class to test, see error message below. It works as expected if run using maven (Run As / Maven test).
See: https://github.com/olovm/modular2.git for an example project

I can work around the problem by adding the following to the automatically created run config in the Arguments tab, under VM Options:
--add-opens org.example.modular2/org.example.modular2=ALL-UNNAMED
but it needs to be done for every unique test run :(

This is when running on eclipse 201812, latest beta of testng-eclipse and java 11.


### Any relate message in "Error Log" view
[RemoteTestNG] revisions:
	git.commit.id=8139f16
	git.build.version=1.4.0-SNAPSHOT
	git.branch=8139f16eccd87e35f80a8bf2edc32dcad7f1194b
[RemoteTestNG] loaded class org.testng.internal.Version at file:/home/olov/.m2/repository/org/testng/testng/6.14.3/testng-6.14.3.jar
[RemoteTestNG] detected TestNG version 6.14.3
[RemoteTestNG] Invoked with -serport 36573 -protocol json -d /home/olov/workspace/modular2/test-output /tmp/testng-eclipse-1543958007/testng-customsuite.xml 
[BaseMessageSender] Waiting for Eclipse client on localhost:36573
[BaseMessageSender] Received a connection from Eclipse on localhost:36573
[BaseMessageSender] Connection established, starting reader thread
[BaseMessageSender] ReaderThread waiting for an admin message
[JsonMessageSender] Sending message [GenericMessage ==> suiteCount:1, testCount:1]
[BaseMessageSender] ReaderThread received admin message:>ACK
[BaseMessageSender] Admin message:>ACK
[BaseMessageSender] Received ACK:>ACK
[BaseMessageSender] Admin message:>ACK
[BaseMessageSender] Received ACK:>ACK
org.testng.TestNGException: 
Cannot instantiate class org.example.modular2.Modular2Test
	at org.testng.internal.ObjectFactoryImpl.newInstance(ObjectFactoryImpl.java:30)
	at org.testng.internal.ClassHelper.createInstance1(ClassHelper.java:423)
	at org.testng.internal.ClassHelper.createInstance(ClassHelper.java:336)
	at org.testng.internal.ClassImpl.getDefaultInstance(ClassImpl.java:125)
	at org.testng.internal.ClassImpl.getInstances(ClassImpl.java:190)
	at org.testng.TestClass.getInstances(TestClass.java:95)
	at org.testng.TestClass.initTestClassesAndInstances(TestClass.java:81)
	at org.testng.TestClass.init(TestClass.java:73)
	at org.testng.TestClass.<init>(TestClass.java:38)
	at org.testng.TestRunner.initMethods(TestRunner.java:389)
	at org.testng.TestRunner.init(TestRunner.java:271)
	at org.testng.TestRunner.init(TestRunner.java:241)
	at org.testng.TestRunner.<init>(TestRunner.java:192)
	at org.testng.remote.support.RemoteTestNG6_12$1.newTestRunner(RemoteTestNG6_12.java:33)
	at org.testng.remote.support.RemoteTestNG6_12$DelegatingTestRunnerFactory.newTestRunner(RemoteTestNG6_12.java:66)
	at org.testng.SuiteRunner$ProxyTestRunnerFactory.newTestRunner(SuiteRunner.java:713)
	at org.testng.SuiteRunner.init(SuiteRunner.java:260)
	at org.testng.SuiteRunner.<init>(SuiteRunner.java:198)
	at org.testng.TestNG.createSuiteRunner(TestNG.java:1295)
	at org.testng.TestNG.createSuiteRunners(TestNG.java:1273)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1128)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:115)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:251)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:77)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make public org.example.modular2.Modular2Test() accessible: module org.example.modular2 does not "exports org.example.modular2" to unnamed module @7530ad9c
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:340)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:280)
	at java.base/java.lang.reflect.Constructor.checkCanSetAccessible(Constructor.java:189)
	at java.base/java.lang.reflect.Constructor.setAccessible(Constructor.java:182)
	at org.testng.internal.ObjectFactoryImpl.newInstance(ObjectFactoryImpl.java:23)
	... 25 more
java.net.SocketException: Socket closed
	at java.base/java.net.SocketInputStream.socketRead0(Native Method)
	at java.base/java.net.SocketInputStream.socketRead(SocketInputStream.java:115)
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:168)
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:140)


### The Dependency Management tool for your project

- [x] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse Buildpath (aka. Use "TestNG Library" for your project in Eclipse)

### Operating System

- [ ] Windows
- [x] Linux
- [ ] OSX

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -17,7 +17,7 @@ Supported Metrics:
 * Fixed #407: eclipse plugin generates invalid java for toString & hashCode (@dabraham02124 & Nick Tan)
 * Fixed #408: Doesn't build against or run on Eclipse 4.10 (2018-12)
 * Fixed #410: After adding module-info.java, testng-eclipse no longer finds test classes in maven project.
-
+* Fixed #413: Testng-eclipse can not find classes in non exported packages in a java modular maven project.
 
 ## 6.14
 

--- a/testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestNGLaunchConfigurationDelegate.java
+++ b/testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestNGLaunchConfigurationDelegate.java
@@ -7,24 +7,18 @@ import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
-import org.eclipse.core.resources.IFolder;
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.debug.core.DebugPlugin;
 import org.eclipse.debug.core.ILaunch;
 import org.eclipse.debug.core.ILaunchConfiguration;
-import org.eclipse.jdt.core.IJavaElement;
 import org.eclipse.jdt.core.IJavaProject;
-import org.eclipse.jdt.core.IPackageFragment;
-import org.eclipse.jdt.core.IPackageFragmentRoot;
-import org.eclipse.jdt.core.JavaModelException;
 import org.eclipse.jdt.launching.AbstractJavaLaunchConfigurationDelegate;
 import org.eclipse.jdt.launching.AbstractVMInstall;
 import org.eclipse.jdt.launching.ExecutionArguments;
@@ -159,26 +153,16 @@ public class TestNGLaunchConfigurationDelegate
     }
 
     // Program & VM args
-    VMRunnerConfiguration runConfig = createVMRunner(configuration, launch,
-        jproject, port, mode);
+    VMRunnerConfiguration runConfig = createVMRunner(configuration, launch, jproject, port, mode);
 
     ExecutionArguments execArgs = new ExecutionArguments(
         ConfigurationHelper.getJvmArgs(configuration), ""); //$NON-NLS-1$
     ArrayList<String> vmArguments= new ArrayList<>();
+    // add the add-opens vm args collected in createVMRunner
+    vmArguments.addAll(Arrays.asList(runConfig.getVMArguments()));
     vmArguments.addAll(Arrays.asList(DebugPlugin.parseArguments(getVMArguments(configuration, mode))));
     vmArguments.addAll(Arrays.asList(execArgs.getVMArgumentsArray()));
     if (JavaRuntime.isModularProject(jproject)) {
-      // issue #413, ideally it's better to open packages contains the tests
-      //            however, TestNG support launch the test from Java method, type, package and testng suite xml file
-      //            it's hard to get the packages used in suite xml file
-      //            the quick and dirty way below, is to open all source packages of this project
-      String sourceModuleName = jproject.getModuleDescription().getElementName();
-      for (String pkgName : getSourcePackages(jproject)) {
-        vmArguments.add("--add-opens");
-        vmArguments.add(sourceModuleName + "/" + pkgName + "=ALL-UNNAMED");
-      }
-      // end #413
-
       vmArguments.add("--add-modules=ALL-MODULE-PATH"); //$NON-NLS-1$
     }
 
@@ -209,29 +193,6 @@ public class TestNGLaunchConfigurationDelegate
     return runConfig;
   }
 
-  /**
-   * get all source package under this project.
-   * 
-   * @param javaProject
-   * @return the String set of the package name
-   * @throws JavaModelException
-   */
-  private Set<String> getSourcePackages(IJavaProject javaProject) throws JavaModelException {
-    Set<String> pkgs = new HashSet<>();
-    for (IPackageFragmentRoot pkgFragmentRoot : javaProject.getPackageFragmentRoots()) {
-      if (!pkgFragmentRoot.isArchive()) {
-        for (final IJavaElement pkg : pkgFragmentRoot.getChildren()) {
-          if (!pkg.getElementName().isEmpty() && !(pkg instanceof IFolder)) {
-            if (((IPackageFragment) pkg).containsJavaResources()) {
-              pkgs.add(((IPackageFragment) pkg).getElementName());
-            }
-          }
-        }
-      }
-    }
-    return pkgs;
-  }
-  
   @Override
   public String getMainTypeName(ILaunchConfiguration configuration)
       throws CoreException {
@@ -350,8 +311,8 @@ public class TestNGLaunchConfigurationDelegate
       argv.add("false");
     }
 
-    List<LaunchSuite> launchSuiteList = ConfigurationHelper
-        .getLaunchSuites(jproject, configuration);
+    List<String> vmArguments= new ArrayList<>();
+    List<LaunchSuite> launchSuiteList = ConfigurationHelper.getLaunchSuites(jproject, configuration, vmArguments);
     List<String> suiteList = new ArrayList<String>();
     List<String> tempSuites = new ArrayList<String>();
 
@@ -380,6 +341,7 @@ public class TestNGLaunchConfigurationDelegate
     }
 
     vmConfig.setProgramArguments(argv.toArray(new String[argv.size()]));
+    vmConfig.setVMArguments(vmArguments.toArray(new String[vmArguments.size()]));
 
     return vmConfig;
   }

--- a/testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestNGLaunchConfigurationDelegate.java
+++ b/testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestNGLaunchConfigurationDelegate.java
@@ -7,18 +7,24 @@ import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
+import org.eclipse.core.resources.IFolder;
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.debug.core.DebugPlugin;
 import org.eclipse.debug.core.ILaunch;
 import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.jdt.core.IJavaElement;
 import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IPackageFragment;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+import org.eclipse.jdt.core.JavaModelException;
 import org.eclipse.jdt.launching.AbstractJavaLaunchConfigurationDelegate;
 import org.eclipse.jdt.launching.AbstractVMInstall;
 import org.eclipse.jdt.launching.ExecutionArguments;
@@ -162,6 +168,17 @@ public class TestNGLaunchConfigurationDelegate
     vmArguments.addAll(Arrays.asList(DebugPlugin.parseArguments(getVMArguments(configuration, mode))));
     vmArguments.addAll(Arrays.asList(execArgs.getVMArgumentsArray()));
     if (JavaRuntime.isModularProject(jproject)) {
+      // issue #413, ideally it's better to open packages contains the tests
+      //            however, TestNG support launch the test from Java method, type, package and testng suite xml file
+      //            it's hard to get the packages used in suite xml file
+      //            the quick and dirty way below, is to open all source packages of this project
+      String sourceModuleName = jproject.getModuleDescription().getElementName();
+      for (String pkgName : getSourcePackages(jproject)) {
+        vmArguments.add("--add-opens");
+        vmArguments.add(sourceModuleName + "/" + pkgName + "=ALL-UNNAMED");
+      }
+      // end #413
+
       vmArguments.add("--add-modules=ALL-MODULE-PATH"); //$NON-NLS-1$
     }
 
@@ -192,6 +209,29 @@ public class TestNGLaunchConfigurationDelegate
     return runConfig;
   }
 
+  /**
+   * get all source package under this project.
+   * 
+   * @param javaProject
+   * @return the String set of the package name
+   * @throws JavaModelException
+   */
+  private Set<String> getSourcePackages(IJavaProject javaProject) throws JavaModelException {
+    Set<String> pkgs = new HashSet<>();
+    for (IPackageFragmentRoot pkgFragmentRoot : javaProject.getPackageFragmentRoots()) {
+      if (!pkgFragmentRoot.isArchive()) {
+        for (final IJavaElement pkg : pkgFragmentRoot.getChildren()) {
+          if (!pkg.getElementName().isEmpty() && !(pkg instanceof IFolder)) {
+            if (((IPackageFragment) pkg).containsJavaResources()) {
+              pkgs.add(((IPackageFragment) pkg).getElementName());
+            }
+          }
+        }
+      }
+    }
+    return pkgs;
+  }
+  
   @Override
   public String getMainTypeName(ILaunchConfiguration configuration)
       throws CoreException {

--- a/testng-eclipse-plugin/src/main/org/testng/eclipse/ui/util/ConfigurationHelper.java
+++ b/testng-eclipse-plugin/src/main/org/testng/eclipse/ui/util/ConfigurationHelper.java
@@ -6,10 +6,13 @@ import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.Map.Entry;
 
+import org.eclipse.core.resources.IFolder;
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.resources.ResourcesPlugin;
 import org.eclipse.core.runtime.CoreException;
@@ -25,10 +28,12 @@ import org.eclipse.jdt.core.IJavaElement;
 import org.eclipse.jdt.core.IJavaProject;
 import org.eclipse.jdt.core.IMethod;
 import org.eclipse.jdt.core.IPackageFragment;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
 import org.eclipse.jdt.core.IType;
 import org.eclipse.jdt.core.JavaCore;
 import org.eclipse.jdt.core.JavaModelException;
 import org.eclipse.jdt.launching.IJavaLaunchConfigurationConstants;
+import org.eclipse.jdt.launching.JavaRuntime;
 import org.eclipse.jface.preference.IPreferenceStore;
 import org.testng.eclipse.TestNGPlugin;
 import org.testng.eclipse.TestNGPluginConstants;
@@ -353,7 +358,8 @@ public class ConfigurationHelper {
    * @throws CoreException 
    */
   public static List<LaunchSuite> getLaunchSuites(IJavaProject ijp,
-      ILaunchConfiguration configuration) throws CoreException {
+      ILaunchConfiguration configuration, List<String> vmArguments) throws CoreException {
+    String addOpensTargets = JavaRuntime.isModularProject(ijp) ? "ALL-UNNAMED" : null;
     LaunchType type = ConfigurationHelper.getType(configuration);
 
     List<String> packages= null;
@@ -362,6 +368,11 @@ public class ConfigurationHelper {
     Map<String, List<String>> classMethods= null;
 
     if (type == LaunchType.SUITE) {
+      // #413, when launch suite, it's hard to get the test classes/packages, 
+      // open all packages for suite mode
+      for (String pkg : getSourcePackages(ijp)) {
+        collectAddOpensVmArgs(addOpensTargets, vmArguments, pkg, configuration);
+      }
       return createLaunchSuites(ijp.getProject(), getSuites(configuration));
     }
 
@@ -390,6 +401,7 @@ public class ConfigurationHelper {
         for (IPackageFragment pf : ijp.getPackageFragments()) {
           if (pf.getElementName().equals(pkg)) {
             je.add(pf);
+            collectAddOpensVmArgs(addOpensTargets, vmArguments, pkg, configuration);
             break;
           }
         }
@@ -399,6 +411,7 @@ public class ConfigurationHelper {
       for (String clz : testClasses) {
         IType t = ijp.findType(clz);
         if (t != null) {
+          collectAddOpensVmArgs(addOpensTargets, vmArguments, t, configuration);
           je.add(t);
         }
       }
@@ -413,6 +426,7 @@ public class ConfigurationHelper {
               throw new JavaModelException(new CoreException(TestNGPlugin.createError(
                           "no test method [" + mthd + "] was found")));
             }
+            collectAddOpensVmArgs(addOpensTargets, vmArguments, m, configuration);
             je.add(m);
           }
         }
@@ -440,6 +454,82 @@ public class ConfigurationHelper {
                               getLogLevel(configuration),
                               workingDir
            );
+  }
+
+  /**
+   * copied from org.eclipse.jdt.junit.launcher.JUnitLaunchConfigurationDelegate
+   * 
+   * @param addOpensTargets
+   * @param addOpensVmArgs
+   * @param javaElem
+   * @param configuration
+   * @throws CoreException
+   */
+  private static void collectAddOpensVmArgs(String addOpensTargets, List<String> addOpensVmArgs, IJavaElement javaElem, ILaunchConfiguration configuration) throws CoreException {
+    if (addOpensTargets != null) {
+      IPackageFragment pkg= getParentPackageFragment(javaElem);
+      if (pkg != null) {
+        String pkgName= pkg.getElementName();
+        collectAddOpensVmArgs(addOpensTargets, addOpensVmArgs, pkgName, configuration);
+      }
+    }
+  }
+
+  /**
+   * copied from org.eclipse.jdt.junit.launcher.JUnitLaunchConfigurationDelegate
+   * 
+   * @param addOpensTargets
+   * @param addOpensVmArgs
+   * @param pkgName
+   * @param configuration
+   * @throws CoreException
+   */
+  private static void collectAddOpensVmArgs(String addOpensTargets, List<String> addOpensVmArgs, String pkgName, ILaunchConfiguration configuration) throws CoreException {
+    if (addOpensTargets != null) {
+      IJavaProject javaProject= getJavaProject(configuration);
+      String sourceModuleName= javaProject.getModuleDescription().getElementName();
+      addOpensVmArgs.add("--add-opens"); //$NON-NLS-1$
+      addOpensVmArgs.add(sourceModuleName + "/" + pkgName + "=" + addOpensTargets); //$NON-NLS-1$ //$NON-NLS-2$     
+    }
+  }
+
+  /**
+   * copied from org.eclipse.jdt.junit.launcher.JUnitLaunchConfigurationDelegate
+   * @param element
+   * @return
+   */
+  private static IPackageFragment getParentPackageFragment(IJavaElement element) {
+    IJavaElement parent= element.getParent();
+    while (parent != null) {
+      if (parent instanceof IPackageFragment) {
+        return (IPackageFragment) parent;
+      }
+      parent= parent.getParent();
+    }
+    return null;
+  }
+
+  /**
+   * get all source package under this project.
+   * 
+   * @param javaProject
+   * @return the String set of the package name
+   * @throws JavaModelException
+   */
+  private static Set<String> getSourcePackages(IJavaProject javaProject) throws JavaModelException {
+    Set<String> pkgs = new HashSet<>();
+    for (IPackageFragmentRoot pkgFragmentRoot : javaProject.getPackageFragmentRoots()) {
+      if (!pkgFragmentRoot.isArchive()) {
+        for (final IJavaElement pkg : pkgFragmentRoot.getChildren()) {
+          if (!pkg.getElementName().isEmpty() && !(pkg instanceof IFolder)) {
+            if (((IPackageFragment) pkg).containsJavaResources()) {
+              pkgs.add(((IPackageFragment) pkg).getElementName());
+            }
+          }
+        }
+      }
+    }
+    return pkgs;
   }
 
   public static IPath getWorkingDirectoryPath(ILaunchConfiguration configuration) throws CoreException {

